src/inc/til/unicode.h

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+namespace til
+{
+    static constexpr bool is_surrogate(const wchar_t wch) noexcept
+    {
+        return (wch & 0xF800) == 0xD800;
+    }
+
+    static constexpr bool is_leading_surrogate(const wchar_t wch) noexcept
+    {
+        return (wch & 0xFC00) == 0xD800;
+    }
+
+    static constexpr bool is_trailing_surrogate(const wchar_t wch) noexcept
+    {
+        return (wch & 0xFC00) == 0xDC00;
+    }
+
+    constexpr std::wstring_view utf16_next(std::wstring_view wstr) noexcept
+    {
+        auto it = wstr.begin();
+        const auto end = wstr.end();
+        auto p = L"\uFFFD";
+        size_t l = 1;
+
+        if (it != end)
+        {
+            const auto wch = *it;
+            p = &*it;
+
+            if (is_surrogate(wch))
+            {
+                ++it;
+                const auto wch2 = it != end ? *it : wchar_t{};
+                if (is_leading_surrogate(wch) && is_trailing_surrogate(wch2))
+                {
+                    l = 2;
+                    ++it;
+                }
+                else
+                {
+                    p = L"\uFFFD";
+                }
+            }
+        }
+
+        return { p, l };
+    }
+
+    struct utf16_iterator
+    {
+        struct sentinel
+        {
+        };
+
+        struct iterator
+        {
+            using iterator_category = std::forward_iterator_tag;
+            using value_type = std::wstring_view;
+            using reference = value_type&;
+            using pointer = value_type*;
+            using difference_type = std::ptrdiff_t;
+
+            explicit constexpr iterator(utf16_iterator& p) noexcept :
+                _p{ p }
+            {
+            }
+
+            const value_type& operator*() const noexcept
```

Comment:
```suggestion
            const reference operator*() const noexcept
```

Reply:
Unfortunately C++ is an excellent language and `const reference` actually translates to just `reference`. This is because type aliases act like proper value types. For example when you return a `const int` from a function you don't actually force the caller to make their `int` a constant. You can't, because an `int` is a value and `const` has no effect on those. Same thing happens for type aliases: `const reference` is like `const int` (which is identical to `int` / `reference`).